LeaderboardBackend/Authorization/UserTypeAuthorizationHandler.cs

```diff
@@ -34,37 +26,27 @@ UserTypeRequirement requirement
             return;
         }
 
-        Guid? userId = _authService.GetUserIdFromClaims(context.User);
+        GetUserResult res = await _userService.GetUserFromClaims(context.User);
 
-        if (userId is null)
+        res.Switch(user =>
         {
-            context.Fail();
-            return;
-        }
-
-        User? user = _userService.GetUserById(userId.Value).Result;
-
-        if (user is null || !Handle(user, requirement))
-        {
-            // FIXME: Work out how to fail as a ForbiddenResult.
-            context.Fail();
-            return;
-        }
-
-        context.Succeed(requirement);
-
-        return;
+            if (Handle(user, requirement))
+            {
+                context.Succeed(requirement);
+            }
+            else
+            {
+                context.Fail();
+            }
+        }, badCredentials => context.Fail(new(this, "Bad Credentials")), notFound => context.Fail(new(this, "User Not Found")));
```

Comment:
So if I understand this correctly; in all three failure cases, we will ultimately return a 401 to the client, right?

Reply:
Not exactly. In the case where the user simply does not have permission to access the resource, 403 will be returned. Not much point in hiding endpoints with 404 since the code is open source and everyone can see which endpoints exist anyway.